File: src/main/java/quebec/virtualite/security/SecurityUserManager.java
package quebec.virtualite.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;

@Service
public class SecurityUserManager
{
    @Autowired
    private JdbcTemplate jdbcTemplate;

    private PasswordEncoder passwordEncoder;

    @PostConstruct
    public void init()
    {
        passwordEncoder = PasswordEncoderFactories.createDelegatingPasswordEncoder();
    }

    public void defineUser(String username, String password)
    {
        if (doesUserExist(username))
            return;

        jdbcTemplate.update(
            "INSERT INTO users (username, password, enabled)"
            + " VALUES (?, ?, TRUE)",
            username,
            passwordEncoder.encode(password));

        jdbcTemplate.update(
            "INSERT INTO authorities (username, authority)"
            + " VALUES (?, 'ROLE_USER')",
            username);
    }

    private boolean doesUserExist(String username)
    {
        return jdbcTemplate.queryForObject(
            "SELECT COUNT(*) FROM users WHERE username = ?",
            new Object[]{username},
            Integer.class) == 1;
    }
}

File: src/main/java/quebec/virtualite/backend/security/SecurityUserManager.kt
package quebec.virtualite.backend.security

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import javax.annotation.PostConstruct

@Service
class SecurityUserManager
{
    @Autowired
    private lateinit var jdbcTemplate: JdbcTemplate

    private var passwordEncoder: PasswordEncoder? = null

    @PostConstruct
    fun _init()
    {
        passwordEncoder = PasswordEncoderFactories.createDelegatingPasswordEncoder()
    }

    fun defineUser(username: String, password: String?)
    {
        if (doesUserExist(username))
            return

        jdbcTemplate.update("INSERT INTO users (username, password, enabled)"
            + " VALUES (?, ?, TRUE)", username, passwordEncoder!!.encode(password))

        jdbcTemplate.update("INSERT INTO authorities (username, authority)"
            + " VALUES (?, 'ROLE_USER')", username)
    }

    private fun doesUserExist(username: String): Boolean
    {
        return jdbcTemplate.queryForObject("SELECT COUNT(*) FROM users WHERE username = ?",
            arrayOf<Any>(username), Int::class.java) == 1
    }
}